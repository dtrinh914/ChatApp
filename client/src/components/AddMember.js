import React,{useState} from 'react';
import useInput from '../hooks/useInput';
import MemberResultCard from './MemberResultCard';
import {makeStyles} from '@material-ui/styles';
import Paper from '@material-ui/core/Paper'
import Typography from '@material-ui/core/Typography';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Input from '@material-ui/core/Input';
import Button from '@material-ui/core/Button';
import List from '@material-ui/core/List';
import axios from 'axios';
import uuid from 'uuid/v4';


const useStyles = makeStyles({
    root:{
        display: 'flex',
        position: 'fixed',
        top: 0,
        left: 0,
        zIndex: 500,
        height: '100%',
        width: '100%',
        justifyContent: 'center',
        alignItems: 'center',
        background: 'rgba(0,0,0,0.4)'
    },
    paper:{
        padding: '15px',
        textAlign: 'center',
        background: 'white',
        height: '90%',
        maxHeight: '700px',
        width: '80%',
        maxWidth: '300px',
        overflow: 'auto'
    },
});

export default function AddMember({sendInvite, filterResults, closeAddMem}) {
    const classes = useStyles();
    const [username, setUsername] = useInput();
    const [searchResults, setSearchResults] = useState([]);

    const searchUser = (name) =>{
        const url = '/api/users/search/'+ name;
        axios.get(url, {withCredentials:true})
        .then( res => {
            if(res.data.status === 1){
                setSearchResults(res.data.data);
            }
        })
    }
    
    const handleSubmit = (e) => {
        e.preventDefault();
        searchUser(username);
    }

    return (
        <div className={classes.root}>
            <ClickAwayListener onClickAway={closeAddMem}>
                <Paper className={classes.paper} >
                    <Typography>Add a Member</Typography>
                    <form onSubmit={handleSubmit}>
                        <Input className={classes.input} id='username' name='username' placeholder='Username' 
                        value={username} onChange={setUsername} />
                        <List>
                            {filterResults(searchResults)
                                .map(result => <MemberResultCard key={uuid()} 
                                                                username={result.username} 
                                                                userId={result._id}
                                                                status= {result.status} 
                                                                sendInvite = {sendInvite}
                                                                />)}
                        </List>
                        <Button className={classes.button} type='submit'>Search</Button>
                    </form>
                    <Button className={classes.button} onClick={closeAddMem} >Close</Button>
                </Paper>
            </ClickAwayListener>
        </div>
    )
}
