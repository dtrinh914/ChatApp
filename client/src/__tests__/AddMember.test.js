import React from 'react';
import AddMember from '../components/AddMember';
import {cleanup, fireEvent, render, waitForElement} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import axiosMock from 'axios';

afterEach(cleanup);

jest.mock('axios');

const mockFilterResults = (data) => {
    return data;
}
const mockCloseAddMem = () =>{
    console.log('Clicked Away');
}

test('should match snapshot', ()=>{
    const {container} = render(<AddMember filterResults={mockFilterResults} 
                                    closeAddMem={mockCloseAddMem} />);
    expect(container).toMatchSnapshot();
});

test('should fetch data and return list of users', async () => {
    let {getByTestId, getAllByTestId} = render(<AddMember filterResults={mockFilterResults} 
                                    closeAddMem={mockCloseAddMem} />);
    const addMemberInput = getByTestId('addmember-input');
    const addMemberSubmitButton = getByTestId('addmember-submit-button');

    //user enters text into search input
    fireEvent.change(addMemberInput, {target: {value: 'test_user'}});
    expect(addMemberInput.value).toBe('test_user');

    const mockData = [
                        {_id: "5e28c5c185ffc434b337c266", username: "test_user1", status:'active'},
                        {_id: "5e28c5bc85ffc434b337c264", username: "test_user2", status:'pending'},
                        {_id: "5e28c5c485ffc434b337c268", username: "test_user3", status:'add'}
                     ];
    axiosMock.get.mockResolvedValueOnce({data: {data:mockData, status:1}});

    //user submits entry
    fireEvent.click(addMemberSubmitButton);

    const members = await waitForElement(()=> getAllByTestId('member-result-card'));

    //check to see if data is properly fetched
    expect(axiosMock.get).toHaveBeenCalledTimes(1);
    expect(axiosMock.get).toHaveBeenCalledWith('/api/users/search/test_user', {withCredentials:true});

    //check to see that result contains three items
    expect(members.length).toBe(3);
});