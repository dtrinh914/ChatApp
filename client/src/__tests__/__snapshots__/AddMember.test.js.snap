// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiPaper-root makeStyles-paper-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Add a Member
      </p>
      <form>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline"
        >
          <input
            class="MuiInputBase-input MuiInput-input"
            data-testid="addmember-input"
            id="username"
            name="username"
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
        <ul
          class="MuiList-root MuiList-padding"
        />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          data-testid="addmember-submit-button"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Search
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Close
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
